File: CarlaControl/src/carlacontrol/SolarPositionModel.py
"""The NOAA solar-position algorithm the engine evaluates, in Python.

`USunPositionFunctionLibrary::GetSunPosition`
(`Engine/Plugins/Runtime/SunPosition/Source/SunPosition/Private/SunPosition.cpp:11-149`) is the NOAA
solar-position spreadsheet in pure scalar double arithmetic.
`ACesiumSunSky::UpdateSun_Implementation` (`CesiumSunSky.cpp:405-467`) calls it with the
georeference's latitude and longitude plus the sun's own `TimeZone`, `UseDaylightSavingTime`, `Year`,
`Month`, `Day` and the hour/minute/second derived from `SolarTime`, then stores
`Elevation = sunPosition.Elevation - 180` and `CorrectedElevation = sunPosition.CorrectedElevation
- 180`.

`get_solar_state` reports `Elevation`, the **geometric** angle. The sun's directional light is rotated
by `CorrectedElevation`, the angle with atmospheric refraction applied. Near the horizon the two
differ by roughly half a degree, which is a large fraction of a low sun's elevation, so this class
returns both and the caller chooses.

This is a model of what the engine computes, not a reading of it. `SolarAudit` is what reads the
engine and compares the two.
"""
from __future__ import annotations

import datetime as dt
import math
from dataclasses import dataclass

# ACesiumSunSky::GetHMSFromSolarTime truncates the clock to whole seconds, so a solar time that is
# not a whole number of seconds is evaluated at the second below it. One second of clock is 15
# arcseconds of hour angle, which is at most 0.0042 degrees of elevation at any latitude, and the
# angles come back over the wire as single-precision floats quantised at roughly 1e-5 degrees near
# 180 (the value the engine subtracts from). Agreement closer than the sum of those two is not
# resolvable, so a comparison against this model is meaningless below it.
CLOCK_TRUNCATION_DEGREES = 0.0042
FLOAT_QUANTISATION_DEGREES = 1.6e-5
RESOLUTION_FLOOR_DEGREES = 0.01


@dataclass(frozen=True)
class SunPosition:
    """Where the model puts the sun: both elevations, in degrees, and the azimuth."""

    elevation_deg: float
    corrected_elevation_deg: float
    azimuth_deg: float


class SolarPositionModel:
    """Evaluates the engine's solar algorithm without the engine."""

    @staticmethod
    def julian_day(year: int, month: int, day: int, hour: int, minute: int, second: int) -> float:
        """`FDateTime::GetJulianDay`: 1721425.5 plus whole days since 0001-01-01 plus the day
        fraction."""
        whole = (dt.date(year, month, day) - dt.date(1, 1, 1)).days
        fraction = (hour * 3600 + minute * 60 + second) / 86400.0
        return 1721425.5 + whole + fraction

    @staticmethod
    def hms_from_solar_time(solar_time: float) -> tuple[int, int, int]:
        """`ACesiumSunSky::GetHMSFromSolarTime` (`CesiumSunSky.cpp:575-585`), truncation included."""
        hour = int(solar_time) % 24
        minute = int((solar_time - hour) * 60) % 60
        second = round((solar_time - hour - minute // 60) * 3600) % 60
        return hour, minute, second

    @staticmethod
    def estimate_time_zone_for_longitude(longitude: float) -> float:
        """`ACesiumSunSky::EstimateTimeZoneForLongitude` (`CesiumSunSky.cpp:570-573`).

        This is the zone a generated world's sun is spawned with, and it is local mean solar time,
        not the site's civil offset.
        """
        return longitude / 15.0

    @classmethod
    def sun_position(cls, latitude: float, longitude: float, time_zone: float, is_dst: bool,
                     year: int, month: int, day: int,
                     hour: int, minute: int, second: int) -> SunPosition:
        """The sun's position for a clock reading in the given zone."""
        time_offset = time_zone + (1.0 if is_dst else 0.0)
        latitude_rad = math.radians(latitude)

        julian_day = cls.julian_day(year, month, day, hour, minute, second)
        centuries = (julian_day - 2451545.0) / 36525.0

        mean_longitude = math.fmod(
            280.46646 + centuries * (36000.76983 + centuries * 0.0003032), 360.0)
        mean_longitude_rad = math.radians(mean_longitude)
        mean_anomaly = 357.52911 + centuries * (35999.05029 - 0.0001537 * centuries)
        mean_anomaly_rad = math.radians(mean_anomaly)
        eccentricity = 0.016708634 - centuries * (0.000042037 + 0.0000001267 * centuries)

        equation_of_centre = (
            math.sin(mean_anomaly_rad) * (1.914602 - centuries * (0.004817 + 0.000014 * centuries))
            + math.sin(2 * mean_anomaly_rad) * (0.019993 - 0.000101 * centuries)
            + math.sin(3 * mean_anomaly_rad) * 0.000289)
        true_longitude = mean_longitude + equation_of_centre
        apparent_longitude = (true_longitude - 0.00569
                              - 0.00478 * math.sin(math.radians(125.04 - 1934.136 * centuries)))
        apparent_longitude_rad = math.radians(apparent_longitude)

        mean_obliquity = 23.0 + (26.0 + (21.448 - centuries * (
            46.815 + centuries * (0.00059 - centuries * 0.001813))) / 60.0) / 60.0
        obliquity = mean_obliquity + 0.00256 * math.cos(math.radians(125.04 - 1934.136 * centuries))
        obliquity_rad = math.radians(obliquity)
        declination_rad = math.asin(math.sin(obliquity_rad) * math.sin(apparent_longitude_rad))

        var_y = math.tan(obliquity_rad / 2.0) ** 2
        equation_of_time = 4.0 * math.degrees(
            var_y * math.sin(2 * mean_longitude_rad)
            - 2.0 * eccentricity * math.sin(mean_anomaly_rad)
            + 4.0 * eccentricity * var_y * math.sin(mean_anomaly_rad)
            * math.cos(2 * mean_longitude_rad)
            - 0.5 * var_y * var_y * math.sin(4 * mean_longitude_rad)
            - 1.25 * eccentricity * eccentricity * math.sin(2 * mean_anomaly_rad))

        true_solar_time = math.fmod(
            (hour * 60 + minute + second / 60.0) + equation_of_time + 4.0 * longitude
            - 60.0 * time_offset, 1440.0)
        hour_angle = (true_solar_time / 4.0 + 180.0 if true_solar_time < 0
                      else true_solar_time / 4.0 - 180.0)
        hour_angle_rad = math.radians(hour_angle)

        zenith_rad = math.acos(
            math.sin(latitude_rad) * math.sin(declination_rad)
            + math.cos(latitude_rad) * math.cos(declination_rad) * math.cos(hour_angle_rad))
        elevation = 90.0 - math.degrees(zenith_rad)

        tan_elevation = math.tan(math.radians(elevation))
        refraction = 0.0
        if elevation <= 85.0:
            if elevation > 5.0:
                refraction = (58.1 / tan_elevation - 0.07 / tan_elevation ** 3
                              + 0.000086 / tan_elevation ** 5 / 3600.0)
            elif elevation > -0.575:
                refraction = 1735.0 + elevation * (
                    -518.2 + elevation * (103.4 + elevation * (-12.79 + elevation * 0.711)))
            else:
                refraction = -20.772 / tan_elevation
            refraction /= 3600.0

        azimuth_base = math.degrees(math.acos(
            ((math.sin(latitude_rad) * math.cos(zenith_rad)) - math.sin(declination_rad))
            / (math.cos(latitude_rad) * math.sin(zenith_rad))))
        azimuth = (math.fmod(azimuth_base + 180.0, 360.0) if hour_angle > 0
                   else math.fmod(540.0 - azimuth_base, 360.0))

        return SunPosition(elevation, elevation + refraction, azimuth)

    @classmethod
    def at_solar_time(cls, latitude: float, longitude: float, time_zone: float, is_dst: bool,
                      year: int, month: int, day: int, solar_time: float) -> SunPosition:
        """The sun's position for a clock given as a fractional hour, truncated as the engine does."""
        hour, minute, second = cls.hms_from_solar_time(solar_time)
        return cls.sun_position(latitude, longitude, time_zone, is_dst,
                                year, month, day, hour, minute, second)

    @staticmethod
    def shadow_length_ratio(elevation_deg: float) -> float | None:
        """Shadow length per unit object height, `cot(elevation)`. None when the sun is down."""
        if elevation_deg <= 0.0:
            return None
        return 1.0 / math.tan(math.radians(elevation_deg))

File: CarlaNet/src/CarlaNet.CoSim/SolarPositionModel.cs
namespace CarlaNet.CoSim;

/// <summary>
/// The solar-position algorithm the engine evaluates, evaluated without the engine.
/// </summary>
/// <remarks>
/// <para><c>USunPositionFunctionLibrary::GetSunPosition</c>
/// (<c>Engine/Plugins/Runtime/SunPosition/Source/SunPosition/Private/SunPosition.cpp</c>) is the
/// NOAA solar-position spreadsheet. <c>ACesiumSunSky::UpdateSun_Implementation</c> calls it with the
/// georeference origin, its own time zone and date, and the hour, minute and second it decomposes
/// its clock into, and stores the elevation, the refraction-corrected elevation and the azimuth it
/// answers with. This is a port of both halves, precision included: the engine takes latitude,
/// longitude and zone as single-precision floats, converts the latitude to radians in single
/// precision, and hands its angles back as floats offset by 180 degrees. Measured against a running
/// server on 2026-09-21 at two sites and twelve instants, the Python port of the same algorithm
/// agreed to 6.9e-6 degrees, three orders of magnitude inside <see cref="ResolutionFloorDegrees"/>.</para>
///
/// <para><b>The clock decomposition is the engine's, defect included.</b> The engine rounds its clock
/// to whole seconds, and a clock in the last half-second of a minute rounds to sixty seconds without
/// carrying the minute: 07:00:59.6 is evaluated as 07:00:00, a minute early, which is up to 0.25
/// degrees of hour angle. <see cref="AtEngineClock"/> reproduces that, because a comparison against
/// the engine has to be a comparison against what the engine computes. <see cref="AtInstant"/>
/// evaluates the same algorithm at the instant itself, so the difference between the two is what the
/// engine's decomposition costs at that instant -- a property of the engine given a clock, computed
/// rather than observed.</para>
/// </remarks>
public static class SolarPositionModel
{
    /// <summary>
    /// The finest agreement with the engine that is resolvable: the engine's whole-second clock is
    /// worth up to 0.0042 degrees of elevation, and its angles are single-precision floats near 180
    /// degrees, worth about 1.6e-5 degrees. Their sum, rounded up.
    /// </summary>
    public const double ResolutionFloorDegrees = 0.01;

    private const float SinglePrecisionPi = 3.1415926535897932f;

    /// <summary>
    /// The hour, minute and second <c>ACesiumSunSky::GetHMSFromSolarTime</c> decomposes a clock
    /// into -- including the minute it drops when the seconds round up to sixty.
    /// </summary>
    public static (int Hour, int Minute, int Second) EngineClock(double solarTimeHours)
    {
        int hour = (int)Math.Truncate(solarTimeHours) % 24;
        int minute = (int)Math.Truncate((solarTimeHours - hour) * 60.0) % 60;

        // The engine subtracts Minute / 60 in integer arithmetic, which is zero, so this is the
        // seconds into the hour rounded and taken modulo sixty.
        int second = (int)Math.Floor(((solarTimeHours - hour) * 3600.0) + 0.5) % 60;
        return (hour, minute, second);
    }

    /// <summary>
    /// What the engine's sun reports for a clock reading: the clock decomposed as the engine
    /// decomposes it, then evaluated.
    /// </summary>
    public static SunPosition AtEngineClock(double latitude, double longitude, double timeZoneHours,
                                            int year, int month, int day, double solarTimeHours)
    {
        (int hour, int minute, int second) = EngineClock(solarTimeHours);
        return At(latitude, longitude, timeZoneHours, year, month, day, hour, minute, second);
    }

    /// <summary>
    /// The sun at a date and whole-second clock in a zone, as the engine computes it; for a date or
    /// time that does not exist, the sentinel the engine is left holding: -180 degrees of elevation.
    /// </summary>
    public static SunPosition At(double latitude, double longitude, double timeZoneHours,
                                 int year, int month, int day, int hour, int minute, int second)
    {
        if (year < 1 || year > 9999 || month < 1 || month > 12 || day < 1
            || day > DateTime.DaysInMonth(year, month) || hour < 0 || hour > 23 || minute < 0
            || minute > 59 || second < 0 || second > 59)
        {
            // GetSunPosition returns early leaving its output at zero, and the sun then subtracts 180.
            return new SunPosition(-180.0, -180.0, 0.0);
        }

        var instant = new DateTime(year, month, day, hour, minute, second, DateTimeKind.Unspecified);
        return Evaluate(latitude, longitude, timeZoneHours, instant);
    }

    /// <summary>
    /// The sun at an instant in a zone, fractional seconds and all: the same algorithm with no
    /// clock decomposition in front of it.
    /// </summary>
    public static SunPosition AtInstant(double latitude, double longitude, double timeZoneHours,
                                        DateTime localInstant) =>
        Evaluate(latitude, longitude, timeZoneHours, localInstant);

    /// <summary>
    /// The angle between two suns, in degrees: how far apart the two directions the light would
    /// come from are, whatever the azimuth does near the zenith.
    /// </summary>
    public static double SeparationDegrees(double elevationA, double azimuthA,
                                           double elevationB, double azimuthB)
    {
        (double ax, double ay, double az) = Direction(elevationA, azimuthA);
        (double bx, double by, double bz) = Direction(elevationB, azimuthB);
        double crossX = (ay * bz) - (az * by);
        double crossY = (az * bx) - (ax * bz);
        double crossZ = (ax * by) - (ay * bx);
        double cross = Math.Sqrt((crossX * crossX) + (crossY * crossY) + (crossZ * crossZ));
        double dot = (ax * bx) + (ay * by) + (az * bz);
        return Degrees(Math.Atan2(cross, dot));
    }

    private static SunPosition Evaluate(double latitude, double longitude, double timeZoneHours,
                                        DateTime local)
    {
        // The engine's inputs are single precision, and it converts the latitude in single
        // precision too.
        float latitudeSingle = (float)latitude;
        float longitudeSingle = (float)longitude;
        float timeOffset = (float)timeZoneHours;
        double latitudeRad = latitudeSingle * (SinglePrecisionPi / 180f);

        double julianDay = 1721425.5 + (double)(local.Ticks / TimeSpan.TicksPerDay)
                           + local.TimeOfDay.TotalDays;
        double julianCentury = (julianDay - 2451545.0) / 36525.0;

        double geomMeanLongSunDeg =
            (280.46646 + (julianCentury * (36000.76983 + (julianCentury * 0.0003032)))) % 360.0;
        double geomMeanLongSunRad = Radians(geomMeanLongSunDeg);
        double geomMeanAnomSunDeg = 357.52911 + (julianCentury * (35999.05029 - (0.0001537 * julianCentury)));
        double geomMeanAnomSunRad = Radians(geomMeanAnomSunDeg);
        double eccentEarthOrbit = 0.016708634 - (julianCentury * (0.000042037 + (0.0000001267 * julianCentury)));

        double sunEqOfCtr =
            (Math.Sin(geomMeanAnomSunRad) * (1.914602 - (julianCentury * (0.004817 + (0.000014 * julianCentury)))))
            + (Math.Sin(2.0 * geomMeanAnomSunRad) * (0.019993 - (0.000101 * julianCentury)))
            + (Math.Sin(3.0 * geomMeanAnomSunRad) * 0.000289);
        double sunTrueLongDeg = geomMeanLongSunDeg + sunEqOfCtr;
        double sunAppLongDeg = sunTrueLongDeg - 0.00569
                               - (0.00478 * Math.Sin(Radians(125.04 - (1934.136 * julianCentury))));
        double sunAppLongRad = Radians(sunAppLongDeg);

        double meanObliqEclipticDeg = 23.0 + ((26.0 + ((21.448 - (julianCentury
            * (46.815 + (julianCentury * (0.00059 - (julianCentury * 0.001813)))))) / 60.0)) / 60.0);
        double obliqCorrDeg = meanObliqEclipticDeg
                              + (0.00256 * Math.Cos(Radians(125.04 - (1934.136 * julianCentury))));
        double obliqCorrRad = Radians(obliqCorrDeg);
        double sunDeclinRad = Asin(Math.Sin(obliqCorrRad) * Math.Sin(sunAppLongRad));

        double varY = Math.Pow(Math.Tan(obliqCorrRad / 2.0), 2.0);
        double eqOfTimeMinutes = 4.0 * Degrees(
            (varY * Math.Sin(2.0 * geomMeanLongSunRad))
            - (2.0 * eccentEarthOrbit * Math.Sin(geomMeanAnomSunRad))
            + (4.0 * eccentEarthOrbit * varY * Math.Sin(geomMeanAnomSunRad) * Math.Cos(2.0 * geomMeanLongSunRad))
            - (0.5 * varY * varY * Math.Sin(4.0 * geomMeanLongSunRad))
            - (1.25 * eccentEarthOrbit * eccentEarthOrbit * Math.Sin(2.0 * geomMeanAnomSunRad)));

        double trueSolarTimeMinutes = (local.TimeOfDay.TotalMinutes + eqOfTimeMinutes
                                       + (4.0 * longitudeSingle) - (60.0 * timeOffset)) % 1440.0;
        double hourAngleDeg = trueSolarTimeMinutes < 0
            ? (trueSolarTimeMinutes / 4.0) + 180
            : (trueSolarTimeMinutes / 4.0) - 180.0;
        double hourAngleRad = Radians(hourAngleDeg);

        double solarZenithAngleRad = Acos(
            (Math.Sin(latitudeRad) * Math.Sin(sunDeclinRad))
            + (Math.Cos(latitudeRad) * Math.Cos(sunDeclinRad) * Math.Cos(hourAngleRad)));
        double solarElevationAngleDeg = 90.0 - Degrees(solarZenithAngleRad);
        double tanOfSolarElevationAngle = Math.Tan(Radians(solarElevationAngleDeg));

        // The refraction terms as the engine writes them, operator precedence included.
        double refractionDeg = 0.0;
        if (solarElevationAngleDeg <= 85.0)
        {
            if (solarElevationAngleDeg > 5.0)
            {
                refractionDeg = (58.1 / tanOfSolarElevationAngle)
                                - (0.07 / Math.Pow(tanOfSolarElevationAngle, 3))
                                + (0.000086 / Math.Pow(tanOfSolarElevationAngle, 5) / 3600.0);
            }
            else if (solarElevationAngleDeg > -0.575)
            {
                refractionDeg = 1735.0 + (solarElevationAngleDeg * (-518.2 + (solarElevationAngleDeg
                    * (103.4 + (solarElevationAngleDeg * (-12.79 + (solarElevationAngleDeg * 0.711)))))));
            }
            else
            {
                refractionDeg = -20.772 / tanOfSolarElevationAngle;
            }

            refractionDeg /= 3600.0;
        }

        double correctedDeg = solarElevationAngleDeg + refractionDeg;
        double azimuthBase = Degrees(Acos(
            ((Math.Sin(latitudeRad) * Math.Cos(solarZenithAngleRad)) - Math.Sin(sunDeclinRad))
            / (Math.Cos(latitudeRad) * Math.Sin(solarZenithAngleRad))));
        double azimuthDeg = hourAngleDeg > 0.0
            ? (azimuthBase + 180.0) % 360.0
            : (540.0 - azimuthBase) % 360.0;

        // Handed back as single-precision floats offset by 180, and the sun subtracts the 180 again
        // in single precision.
        float elevation = (float)(180.0f + solarElevationAngleDeg) - 180.0f;
        float corrected = (float)(180.0f + correctedDeg) - 180.0f;
        return new SunPosition(elevation, corrected, (float)azimuthDeg);
    }

    private static (double East, double North, double Up) Direction(double elevationDeg, double azimuthDeg)
    {
        double elevation = Radians(elevationDeg);
        double azimuth = Radians(azimuthDeg);
        return (Math.Cos(elevation) * Math.Sin(azimuth), Math.Cos(elevation) * Math.Cos(azimuth),
                Math.Sin(elevation));
    }

    private static double Radians(double degrees) => degrees * (Math.PI / 180.0);

    private static double Degrees(double radians) => radians * (180.0 / Math.PI);

    private static double Asin(double value) => Math.Asin(Math.Clamp(value, -1.0, 1.0));

    private static double Acos(double value) => Math.Acos(Math.Clamp(value, -1.0, 1.0));
}
